lite2/client.go

```diff
@@ -0,0 +1,419 @@
+// obtain the header & vals from the primary or they are invalid (e.g. trust
+// hash does not match with the one from the header).
+//
+// See all Option(s) for the additional configuration.
+func NewClient(
+	chainID string,
+	trustOptions TrustOptions,
+	primary provider.Provider,
+	trustedStore store.Store,
+	options ...Option) (*Client, error) {
+
+	c := &Client{
+		chainID:          chainID,
+		trustingPeriod:   trustOptions.Period,
+		verificationMode: skipping,
+		trustLevel:       DefaultTrustLevel,
+		primary:          primary,
+		trustedStore:     trustedStore,
+		logger:           log.NewNopLogger(),
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Client) initializeWithTrustOptions(options TrustOptions) error {
+	// 1) Fetch and verify the header.
+	h, err := c.primary.SignedHeader(options.Height)
+	if err != nil {
+		return err
+	}
+
+	// NOTE: Verify func will check if it's expired or not.
+	if err := h.ValidateBasic(c.chainID); err != nil {
+		return errors.Wrap(err, "ValidateBasic failed")
+	}
+
+	if !bytes.Equal(h.Hash(), options.Hash) {
+		return errors.Errorf("expected header's hash %X, but got %X", options.Hash, h.Hash())
+	}
+
+	// 2) Fetch and verify the next vals.
+	vals, err := c.primary.ValidatorSet(options.Height + 1)
+	if err != nil {
+		return err
+	}
+
+	// 3) Persist both of them and continue.
+	return c.updateTrustedHeaderAndVals(h, vals)
+}
+
+// SetLogger sets a logger.
+func (c *Client) SetLogger(l log.Logger) {
+	c.logger = l
+}
+
+// TrustedHeader returns a trusted header at the given height or nil if no such
+// header exist. It returns an error if there are some issues with the trusted
+// store, although that should not happen normally. TODO: mention how many
+// headers will be kept by the light client.
+//
+// 0 - the latest.
+// height must be >= 0.
+func (c *Client) TrustedHeader(height int64) (*types.SignedHeader, error) {
+	if height < 0 {
+		return nil, errors.New("negative height")
+	}
+
+	if height == 0 {
+		var err error
+		height, err = c.LastTrustedHeight()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return c.trustedStore.SignedHeader(height)
+}
+
+// LastTrustedHeight returns a last trusted height.
+func (c *Client) LastTrustedHeight() (int64, error) {
+	return c.trustedStore.LastSignedHeaderHeight()
+}
+
+// ChainID returns the chain ID.
+func (c *Client) ChainID() string {
+	return c.chainID
+}
+
+// VerifyHeaderAtHeight fetches the header and validators at the given height
+// and calls VerifyHeader.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeaderAtHeight(height int64, now time.Time) error {
+	if c.trustedHeader.Height >= height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", height, c.trustedHeader.Height)
+	}
```

Comment:
Is it true that it is "already trusted"? Maybe this is an older header that was never verified., (skipping case). Should we lookup the store instead?
Let's say we start with height A (trusted) and we successfully verify for height Z. So we store headers A and Z.
Now we get a request for height Q with A < Q < Z. 
Is this a valid request?  Or is there a requirement that VerifyHeaderAtHeight() must be called with increasing heights for both sequential and skipping case?
If it is a valid request then we should not trust Q automatically just because we trusted Z. It is possible that a verification of Q against A will fail (e.g. less than 1/3rds common validators between A and Q signed even though there were more than 1/3rds common between A and Z -> Note*)
Also, a proper solution for this "backward" case should check that A is still within trusting period and if not perform backwards-sequential verification (i.e. h, h-1, h-2, etc.) or return error if backwards-sequential is not supported.

So to summarize, if we do allow for random (not necessarily increasing in height) client header verification for skipping case then some changes are needed. If we do not, then maybe clarify this someplace.

Note* - if validators are required to change address when re-bonding then probably this is not an issue.

Reply:
Even in bisection I don't think we should ever end up looking up a header that is older than one we've already decided to trust. We should never start bisection from a 'start header' that's before one we've already decided to trust, so presumably this won't happen (but good question & merits a comment).

Reply:
Ok. We do have (potentially) many trusted headers in the store. Any new bisection call for H could start with the highest trusted header <= H. I think the code can support this with minimal changes if we ever need it. 
But at least, the error message we currently have `height #%d is already trusted..` should be changed.

Reply:
> We do have (potentially) many trusted headers in the store. Any new bisection call for H could start with the highest trusted header <= H.

I think it only makes sense with a different provider. Consider this:

a) light client performs bisection and decides to trust H#100
b) later it receives an evidence of attach (lunatic) and learns that H#100 is fabricated
c) light client exits (we should probably remove all "trusted" headers from DB at this point)
c) operator resets it with new primary source and new trusted header